src/twisted/mail/test/test_smtp.py

```diff
@@ -937,17 +980,74 @@ def test_removeCurrentProtocolWhenClientConnectionFailed(self):
         connection is failed.
         """
         reactor = MemoryReactor()
-        sentDeferred = defer.Deferred()
-        clientFactory = smtp.SMTPSenderFactory(
-            "source@address", "recipient@address",
-            BytesIO(b"message"), sentDeferred)
-        connector = reactor.connectTCP("localhost", 25, clientFactory)
-        clientFactory.buildProtocol(None)
-        clientFactory.clientConnectionFailed(connector,
-                                             error.ConnectionDone("Bye."))
+        clientFactory = self.getFactory()
+        connector = self.connect(reactor, clientFactory)
+        self.assertIsNotNone(clientFactory.currentProtocol)
+
+        clientFactory.clientConnectionFailed(
+            connector, Failure(error.ConnectionDone("Bye.")))
+
         self.assertEqual(clientFactory.currentProtocol, None)
 
 
+    def test_clientConnectionLostConnectionDoneWhenNotFinished(self):
+        """
+        When connection is closed with ConnectionDone before finalizing the
+        send the message and no more retries are left,
+        it will trigger a failure of a dedicated type,
+        to the deferred used to signal the delivery.
+        """
+        reactor = MemoryReactor()
+        sentDeferred = defer.Deferred()
+        sut = self.getFactory(sentDeferred=sentDeferred, retries=0)
+        connector = self.connect(reactor, sut)
+
+        sut.clientConnectionLost(
+            connector, Failure(error.ConnectionDone("Bye.")))
+
+        failure = self.failureResultOf(sentDeferred)
+        self.assertIsInstance(failure.value, smtp.SMTPConnectError)
+        # The delivery deferred is no longer available on the factory.
+        self.assertFalse(hasattr(sut, 'result'))
```

Comment:
Thanks for testing something that wasn't tested before!  Unfortunately, [`hasattr` sometimes lies under Python 2](https://hynek.me/articles/hasattr/).  Please replace this with `getattr`.